feat(pipeline): Add pipeline config section for artifacts.

jtk54 opened the pull request:
Adds a section to pipeline config for declaring expected artifacts from trigger sources, and adds support in pub/sub triggers to select artifacts the trigger can produce.

jtk54 commented:
![image](https://user-images.githubusercontent.com/5456773/30648621-7719504c-9dec-11e7-92c1-3081d666ffbd.png)


![image](https://user-images.githubusercontent.com/5456773/30648630-7c18f458-9dec-11e7-85cf-6bc88017e1b8.png)

danielpeach commented on app/scripts/modules/core/src/pipeline/config/triggers/triggers.html:
```diff
@@ -18,6 +18,21 @@
       </div>
     </div>
   </page-section>
+  <page-section key="artifacts" label="Artifacts" badge="pipeline.expectedArtifacts.length" no-wrapper="true">
```
I'd call this section `Expected Artifacts`.

danielpeach replied:
I'd also consider moving this down on the page (probably below parameters or notifications), especially since we don't have much support for artifacts yet. I'd also feature flag this.

danielpeach commented on app/scripts/modules/core/src/pipeline/config/triggers/triggers.directive.js:
```diff
@@ -33,5 +34,13 @@ module.exports = angular.module('spinnaker.core.pipeline.config.trigger.triggers
       $scope.pipeline.triggers.push(newTrigger);
     };
 
+    this.addArtifact = function() {
```
Use an arrow function here.

danielpeach commented on app/scripts/modules/core/src/pipeline/config/triggers/trigger.html:
```diff
@@ -37,6 +37,46 @@
+                         class="form-control input-sm">
+                <ui-select-match>{{$item.name}}</ui-select-match>
+                <ui-select-choices repeat="artifact in pipeline.expectedArtifacts">
+                  <span ng-bind-html="artifact.name"></span>
```
If the artifact doesn't have a name, I see an empty option here.

jtk54 commented on app/scripts/modules/core/src/pipeline/config/triggers/artifacts/artifact.html:
```diff
@@ -0,0 +1,49 @@
+      <div class="form-group row">
+        <div class="col-md-10">
+          <div class="row">
+            <label class="col-md-3 sm-label-right">Type</label>
```
Added tooltips describing the forms and added the `required` and validation.

jtk54 commented on app/scripts/modules/core/src/pipeline/config/triggers/artifacts/artifact.html:
```diff
@@ -0,0 +1,49 @@
+        </div>
+      </div>
+
+      <div class="form-group row">
```
Added `Ignore` as another option.

lwander commented:
spinnaker/spinnaker#1915

anotherchrisberry commented:
It's probably worth adding some explanatory text to the top of the section. It's not clear to me what these fields are or why I'd need or want to use them.

jtk54 commented:
I fixed up the empty artifact suggestion in the ui-select and added explanatory text to the section with another help field. PTAL @danielpeach and @anotherchrisberry and let me know what you think.

anotherchrisberry commented on app/scripts/modules/core/src/pipeline/config/triggers/artifacts/artifact.component.ts:
```diff
@@ -0,0 +1,41 @@
+      if (found) {
+        t.expectedArtifacts = t.expectedArtifacts.filter(a => !this.isEqual(a, found));
+      }
+    });
```
You don't need `map` since you're not doing anything with it, so should probably use `forEach` (just to be more clear that you're just iterating through the items).

Seems like this could be simplified quite a bit (unless I'm missing something):
```javascript
this.pipeline.triggers
  .forEach(t => t.expectedArtifacts = t.expectedArtifacts.filter(a => !this.isEqual(a, this.artifact)));
```

jtk54 replied:
Yep good call -- there was a conundrum with `isEqual` causing weird behavior since angular was add a `hashCode` field. I had to do some contortions to figure out what was going on and didn't revert it.

anotherchrisberry commented on app/scripts/modules/core/src/pipeline/config/triggers/triggers.html:
```diff
@@ -36,6 +36,27 @@
   <page-section key="parameters" label="Parameters" no-wrapper="true" badge="pipeline.parameterConfig.length">
     <parameters pipeline="pipeline"></parameters>
   </page-section>
+  <page-section key="artifacts" label="Expected Artifacts" badge="pipeline.expectedArtifacts.length" no-wrapper="true">
```
Is this only a thing for pub/sub, or are there other triggers that will include artifacts?

jtk54 replied:
We'll have other triggers include artifacts. Ideally eventually most/all triggers would be able to do this. We have a few starting use cases with pub/sub -> gae and docker -> k8s v2.

jtk54 commented:
Feature flagged this with Dan's help, take a look.

anotherchrisberry commented:
LGTM
